File: src/chatguide/core/task.py
"""Task object representing a unit of work."""

from typing import List, Dict, Any, Optional
from dataclasses import dataclass, field

@dataclass
class Task:
    """Represents a single task in the conversation plan."""
    
    id: str
    description: str
    expects: List[str] = field(default_factory=list)
    tools: List[Dict[str, Any]] = field(default_factory=list)
    silent: bool = False
    
    # Runtime state
    status: str = "pending"  # pending, in_progress, completed
    result: Optional[Dict[str, Any]] = None
    
    def complete(self, key: str, value: Any):
        """Mark task as complete with extracted value."""
        self.status = "completed"
        self.result = {"key": key, "value": value}
        
    def is_completed(self) -> bool:
        """Check if task is completed."""
        return self.status == "completed"
    
    def to_dict(self) -> Dict[str, Any]:
        """Serialize task state."""
        return {
            "id": self.id,
            "description": self.description,
            "expects": self.expects,
            "tools": self.tools,
            "silent": self.silent,
            "status": self.status,
            "result": self.result
        }

File: js/src/core/Task.ts
export interface TaskData {
  id: string;
  description: string;
  expects: string[];
  tools: Array<Record<string, any>>;
  silent: boolean;
  status?: string;
  result?: Record<string, any>;
}

export class Task {
  id: string;
  description: string;
  expects: string[] = [];
  tools: Array<Record<string, any>> = [];
  silent: boolean = false;
  status: 'pending' | 'in_progress' | 'completed' = 'pending';
  result?: Record<string, any>;

  constructor(data: TaskData) {
    this.id = data.id;
    this.description = data.description;
    this.expects = data.expects || [];
    this.tools = data.tools || [];
    this.silent = data.silent || false;
    this.status = (data.status as any) || 'pending';
    this.result = data.result;
  }

  complete(key: string, value: any): void {
    this.status = 'completed';
    this.result = { key, value };
  }

  isCompleted(): boolean {
    return this.status === 'completed';
  }

  toDict(): Record<string, any> {
    return {
      id: this.id,
      description: this.description,
      expects: this.expects,
      tools: this.tools,
      silent: this.silent,
      status: this.status,
      result: this.result,
    };
  }
}

